RealmSwift/Projection.swift

```diff
@@ -0,0 +1,642 @@
+        // wrapper is used on an ObjectBase subclass, which is the only thing we support.
+        get { fatalError("called wrappedValue getter") }
+        // swiftlint:disable:next unused_setter_value
+        set { fatalError("called wrappedValue setter") }
+    }
+
+    /// :nodoc:
+    public static subscript<EnclosingSelf: Projection<T>>(
+        _enclosingInstance observed: EnclosingSelf,
+        wrapped wrappedKeyPath: ReferenceWritableKeyPath<EnclosingSelf, Value>,
+        storage storageKeyPath: ReferenceWritableKeyPath<EnclosingSelf, Self>
+        ) -> Value {
+        get {
+            let storage = observed[keyPath: storageKeyPath]
+            return observed.rootObject[keyPath: storage._projectedKeyPath]
+        }
+        set {
+            guard let keyPath = observed[keyPath: storageKeyPath].projectedKeyPath as? WritableKeyPath<T, Value> else {
+                preconditionFailure("KeyPath is not writable")
+            }
+            observed.rootObject[keyPath: keyPath] = newValue
+        }
+    }
+
+    /// Declares a property which is lazily initialized to the type's default value.
+    public init(_ projectedKeyPath: KeyPath<T, Value>) {
+        self._projectedKeyPath = projectedKeyPath
+    }
+}
+
+// MARK: Projection Schema
+private struct ProjectedMetadata {
+    let projectedKeyPath: AnyKeyPath
+    let originPropertyKeyPathString: String
+    let label: String
+}
+
+private var schema = [ObjectIdentifier: [ProjectedMetadata]]()
+
+// MARK: ProjectionOservable
+/**
+  A type erased Projection.
+ 
+ ProjectionObservable is a Combine publisher
+ */
+public protocol ProjectionObservable: AnyObject {
+    /// The Projection's underlying type - a child of Realm `Object` or `EmbeddedObject`.
+    associatedtype Root: ObjectBase
+    /// The object being projected
+    var rootObject: Root { get }
+    /// :nodoc:
+    init(projecting object: Root)
+}
+
+extension ObjectChange {
+    fileprivate static func processChange(_ objectChange: ObjectChange<T.Root>,
+                                          _ projection: T) -> ObjectChange<T> where T: ProjectionObservable {
+        let schema = projection._schema
+        switch objectChange {
+        case .error(let error):
+            return .error(error)
+        case .change(let object, let objectPropertyChanges):
+            let newProjection = T(projecting: object)
+            let projectedPropertyChanges: [PropertyChange] = objectPropertyChanges.map { propChange in
+                // read the metadata for the property whose origin name matches
+                // the changed property's name
+                let lock = NSLock()
+                lock.lock()
+                let propertyMetadata = schema.first(where: {
+                    $0.originPropertyKeyPathString == propChange.name
+                })!
+                lock.unlock()
+                var change: (name: String?, oldValue: Any?, newValue: Any?) = (nil, nil, nil)
```

Comment:
This is never actually used as a tuple and should probably just be three local variables?

Reply:
fixed